SimulationAgent/SimulationRunner.cs

```diff
@@ -152,146 +147,129 @@ public class SimulationRunner : ISimulationRunner
+            // Start threads
+            this.TryToStartStateThread();
 
-                this.TryToStartConnectionThread();
+            this.TryToStartConnectionThread();
 
-                this.TryToStartTelemetryThreads();
+            this.TryToStartTelemetryThreads();
 
-                this.TryToStartPropertiesThread();
-            }
+            this.TryToStartPropertiesThread();
         }
 
         public void Stop()
         {
-            lock (this.startLock)
-            {
-                if (!this.running) return;
-
-                this.log.Info("Stopping simulation...");
+            if (this.stopping || !this.running) return;
 
-                this.running = false;
+            // Used to prevent contentions between Start and Stop, setting this here prevents another start to be enqueued
+            this.stopping = true;
 
-                foreach (var device in this.deviceConnectionActors)
+            var attempts = 20;
+            while (this.starting)
+            {
+                Thread.Sleep(1000);
+                if (this.starting && attempts-- <= 0)
                 {
-                    device.Value.Stop();
+                    throw new Exception("Unable to stop the simulation, the simulation is also starting");
                 }
+            }
 
-                foreach (var device in this.deviceTelemetryActors)
-                {
-                    device.Value.Stop();
-                }
+            if (this.stopping || !this.running) return;
 
-                foreach (var device in this.devicePropertiesActors)
-                {
-                    device.Value.Stop();
-                }
+            this.log.Info("Stopping simulation...");
+
+            this.running = false;
 
-                // Allow 3 seconds to complete before stopping the threads
-                Thread.Sleep(3000);
-                this.TryToStopStateThread();
-                this.TryToStopConnectionThread();
-                this.TryToStopTelemetryThreads();
-                this.TryToStopPropertiesThread();
+            foreach (var device in this.deviceConnectionActors)
+            {
+                device.Value.Stop();
+            }
 
-                // Reset local state
-                this.deviceStateActors.Clear();
-                this.deviceTelemetryActors.Clear();
-                this.deviceConnectionActors.Clear();
-                this.devicePropertiesActors.Clear();
-                this.starting = false;
+            foreach (var device in this.deviceTelemetryActors)
+            {
+                device.Value.Stop();
+            }
 
-                // Reset rateLimiting counters
-                this.rateLimiting.ResetCounters();
+            foreach (var device in this.devicePropertiesActors)
+            {
+                device.Value.Stop();
             }
+
+            // Allow 3 seconds to complete before stopping the threads
+            Thread.Sleep(3000);
```

Comment:
Should this be a concurrency config setting, or do we have confidence that three seconds will always be enough time to wait here?

Reply:
no need to make it configurable, and this is going away once the simulation context work is done